pkg/redirectpolicy/manager.go

```diff
@@ -672,22 +661,53 @@ func (rpm *Manager) processConfigWithNamedPorts(config *LRPConfig, pods ...*podM
 				}
 			}
 			if len(bes4) > 0 {
-				rpm.upsertServiceWithBackends(config, feM, pod.id, bes4)
+				rpm.upsertFrontendMapping(config, feM, pod.id, bes4)
 			} else if len(bes6) > 0 {
-				rpm.upsertServiceWithBackends(config, feM, pod.id, bes6)
+				rpm.upsertFrontendMapping(config, feM, pod.id, bes6)
 			}
 		}
 	}
 }
 
-// upsertServiceWithBackends updates policy config internal state and upserts
-// service with the given pod backends.
-func (rpm *Manager) upsertServiceWithBackends(config *LRPConfig, frontendMapping *feMapping, podID podID, backends []backend) {
-	frontendMapping.backends = backends
-	rpm.policyPods[podID] = append(rpm.policyPods[podID], podPolicyInfo{
-		policyID: config.id,
-		backends: backends,
-	})
+// upsertFrontendMapping updates policy config internal state and upserts
+// service with the policy frontend and mapped backends.
+func (rpm *Manager) upsertFrontendMapping(config *LRPConfig, frontendMapping *feMapping, podID podID, backends []backend) {
+	newFePods := make([]backend, 0, len(frontendMapping.podBackends))
+	// Update the frontend mapped backends slice, keeping the order same.
+	// TODO (aditi) unit test case
```

Comment:
+1, do you intend to address that in this PR?

Having tests in place before refactoring is a great way to provide reassurance that things are still working like they were before.

Reply:
> If I'm reading it right, if the pod has multiple backends then we'll issue multiple service updates but we should be able to just calculate all backends for a service at once and then update for all known backends.

We keep a map of frontend and its mapped backends. A frontend may have multiple backend pods, each with multiple backends (IP+port tuples). Whenever there is an update for a backend pod, we update all the frontend mappings that are mapped to this backend pod. Before we can override the current service entry for the frontend, we re-compute the backends slice for every updated frontend with the updated pod backends.

Reply:
(for the reference the comment in my original feedback which is quoted above is the topic of the thread below: https://github.com/cilium/cilium/pull/13362/files/c0a9634f8e410df5119d7cccb14dae981cc29b21..f52c83394229c9cb01ca44cc68387cf5597344f9#r499926266)